datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -466,15 +625,122 @@ impl GroupedHashAggregateStream {
         for acc in self.accumulators.iter_mut() {
             match self.mode {
                 AggregateMode::Partial => output.extend(acc.state(emit_to)?),
+                _ if spilling => {
+                    // If spilling, output partial state because the spilled data will be
+                    // merged and re-evaluated later.
+                    output.extend(acc.state(emit_to)?)
+                }
                 AggregateMode::Final
                 | AggregateMode::FinalPartitioned
                 | AggregateMode::Single
                 | AggregateMode::SinglePartitioned => output.push(acc.evaluate(emit_to)?),
             }
         }
 
-        self.update_memory_reservation()?;
-        let batch = RecordBatch::try_new(self.schema(), output)?;
+        // emit reduces the memory usage. Ignore Err from update_memory_reservation. Even if it is
+        // over the target memory size after emission, we can emit again rather than returning Err.
+        let _ = self.update_memory_reservation();
+        let batch = RecordBatch::try_new(schema, output)?;
         Ok(batch)
     }
+
+    /// Optimistically, [`Self::group_aggregate_batch`] allows to exceed the memory target slightly
+    /// (~ 1 [`RecordBatch`]) for simplicity. In such cases, spill the data to disk and clear the
+    /// memory. Currently only [`GroupOrdering::None`] is supported for spilling.
+    fn spill_previous_if_necessary(&mut self, batch: &RecordBatch) -> Result<()> {
+        // TODO: support group_ordering for spilling
+        if self.group_values.len() > 0
+            && batch.num_rows() > 0
+            && matches!(self.group_ordering, GroupOrdering::None)
+            && !matches!(self.mode, AggregateMode::Partial)
+            && !self.spill_state.is_stream_merging
+            && self.update_memory_reservation().is_err()
+        {
+            // Use input batch (Partial mode) schema for spilling because
+            // the spilled data will be merged and re-evaluated later.
+            self.spill_state.spill_schema = batch.schema();
+            self.spill()?;
+            self.clear_shrink(batch);
+        }
+        Ok(())
+    }
+
+    /// Emit all rows, sort them, and store them on disk.
+    fn spill(&mut self) -> Result<()> {
+        let emit = self.emit(EmitTo::All, true)?;
+        let sorted = sort_batch(&emit, &self.spill_state.spill_expr, None)?;
```

Comment:
I might be missing it. But I'm wondering why we need to sort spilling batches? As they are simply partial aggregation result, and we do spilling only for `GroupOrdering::None`, does the order of spilling batches matter?

Reply:
`streaming_merge` expects all batches are pre-sorted (because it is merge sort).

Reply:
`streaming_merge` needs to sort because it is for sort operator. I mean for this hash aggregation, I wonder why it is needed. If it is unnecessary, that adds redundant performance cost on sorting.

Reply:
Using the example of final aggregation in the doc
https://github.com/apache/arrow-datafusion/pull/7400/files#diff-0b72cdc7a5e02b6650430c3eb7bdea4f1d2d32867b927633d92abb4fcf31c81bR187-R202
We have to re-group for rows with `a = 2`. Please notice `a = 2` can be in any spilled batches.

The spills happened because the memory is insufficient for all rows. The only choice we have left is merge based aggregation.

Reply:
Hmm, firstly the spilled data is partial aggregation states, right? This tries to do is to spill them and read back later after all inputs are done (all partial aggregation states are aggregated or spilled). The merged stream of spilled batches are read in as new inputs, so they are continuously aggregated into final aggregation result.

As the partial aggregation inputs should be unordered (this is not sort-based aggregation but hash aggregation), that is I wonder why we need to sort them before spilling. It doesn't matter for rows with `a = 2` in which spilled batches. Once these rows are read, they are aggregated as expected.

Reply:
If the sort can be removed, for now it is there simply because we want to reuse `streaming_merge`, then it is fine. But it's better to leave a TODO to remove the redundant sorting later.

Reply:
> ... so they are continuously aggregated into final aggregation result.

During this re-grouping time, we may run out of memory again. The only option is to streaming read and early outputting. In order to do so, the input for regrouping has to be sorted. Otherwise we cannot output the final aggregation results early and that makes the hash table keep all rows in memory. `a = 2` can be at very first and very last. So the very last step has to be a sort based aggregation in order to avoid OOM.

Reply:
It makes sense finally. This should be the main reason why sorting is necessary here.

Could you add a comment in the code for that?

Reply:
I added a short version of it to the doc section